cpython.proj

```diff
@@ -0,0 +1,178 @@
+      IgnoreStandardErrorWarningFormat="true" />
+    <Exec WorkingDirectory="$(PythonInstallDir)"
+      Command="lipo -thin $(LipoArchitecture) bin/python3 -output bin/python3.$(TargetArchitecture)"
+      IgnoreStandardErrorWarningFormat="true"
+      Condition="'$(BuildOS)' == 'OSX'" />
+    <Exec WorkingDirectory="$(PythonInstallDir)"
+      Command="lipo -thin $(LipoArchitecture) bin/python$(PythonMajorVersion) -output bin/python$(PythonMajorVersion).$(TargetArchitecture)"
+      IgnoreStandardErrorWarningFormat="true"
+      Condition="'$(BuildOS)' == 'OSX'" />
+    <Move Condition="'$(BuildOS)' == 'OSX'" SourceFiles="$(PythonInstallDir)bin\python$(PythonMajorVersion).$(TargetArchitecture)" DestinationFiles="$(PythonInstallDir)bin\python$(PythonMajorVersion)" />
+    <Move Condition="'$(BuildOS)' == 'OSX'" SourceFiles="$(PythonInstallDir)bin\python3.$(TargetArchitecture)" DestinationFiles="$(PythonInstallDir)bin\python3" />
+    <Delete Files="$(PythonInstallDir)lib\libpython$(PythonMajorVersion).a" />
+    <ItemGroup>
+      <PyCaches Include="$(PythonInstallDir)**\*.pyc" />
+      <PyCacheDirs Include="@(PyCaches->'%(RelativeDir)')" />
+      <PyCacheDirsDeduplicated Include="@(PyCacheDirs->Distinct())" />
+    </ItemGroup>
+    <RemoveDir Directories="$(PythonInstallDir)include;$(PythonInstallDir)lib\python$(PythonMajorVersion)\test\;@(PyCacheDirsDeduplicated)" />
+  </Target>
+
+  <Target Name="BuildPythonWindows" BeforeTargets="Build" Condition="'$(BuildOS)' == 'Windows_NT'">
+    <PropertyGroup>
+      <PythonHostArch Condition="'$(TargetArchitecture)' == 'x64'">x64</PythonHostArch>
+      <PythonHostArch Condition="'$(TargetArchitecture)' == 'x86'">Win32</PythonHostArch>
+      <PythonHostArch Condition="'$(TargetArchitecture)' == 'arm'">ARM</PythonHostArch>
+      <PythonHostArch Condition="'$(TargetArchitecture)' == 'arm64'">arm64</PythonHostArch>
+    </PropertyGroup>
+    <MakeDir Directories="$(PythonBuildDir)" Condition="!Exists('$(PythonBuildDir)')" />
+    <Exec WorkingDirectory="$(ProjectDir)PCbuild"
+      Command="build.bat -vv -c $(Configuration) -p $(PythonHostArch) -t Build --no-tkinter"
+      IgnoreStandardErrorWarningFormat="true"
+      EnvironmentVariables="PlatformToolset=v142;Py_OutDir=$(PythonBuildDir)"/>
+  </Target>
+
+  <Target Name="LayoutPythonWindows" AfterTargets="BuildPythonWindows"  Condition="'$(BuildOS)' == 'Windows_NT'">
+    <PropertyGroup>
+      <PythonLayoutArch Condition="'$(TargetArchitecture)' == 'x64'">amd64</PythonLayoutArch>
+      <PythonLayoutArch Condition="'$(TargetArchitecture)' == 'x86'">win32</PythonLayoutArch>
+      <PythonLayoutArch Condition="'$(TargetArchitecture)' == 'arm'">arm32</PythonLayoutArch>
+      <PythonLayoutArch Condition="'$(TargetArchitecture)' == 'arm64'">arm64</PythonLayoutArch>
+      <PythonDebugFlag Condition="'$(Configuration)' == 'Debug'">-d</PythonDebugFlag>
+    </PropertyGroup>
+    <Exec WorkingDirectory="C:\"
```

Comment:
why is the working directory necessary?

instead of hardcoding C: we could use `$([System.IO.Path]::GetPathRoot($(MSBuildProjectFullPath)))`

Reply:
This is the one thing it absolutely must not be. During the build, Python emits a python.bat into $(MSBuildProjectFullPath) which points to the just-built python.exe in the build tree. This fails the build during cross-compilation of arm64 python on x64